### PR TITLE
[TWAP]: Add safety check regarding the number of records in updateRecords

### PR DESCRIPTION
**Pull request by stackman27**
Closes: #2644

## Brief Changelog

*(for example:)*
 
  - *The metadata is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *Daemons retrieve the RPC data from the blob cache*

## Testing and Verifying

n/a

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: x/twap/logic.go [hunk 1]
```diff
@@ -103,6 +103,19 @@ func (k Keeper) updateRecords(ctx sdk.Context, poolId uint64) error {
...
 	}
 
+	denoms, err := k.ammkeeper.GetPoolDenoms(ctx, poolId)
+	if err != nil {
+		return err
+	}
+
+	// check if there is k * (k - 1) / 2 records, k = denoms
+	denomNum := len(denoms)
+	expectedRecordsLength := denomNum * (denomNum - 1) / 2
```
Comment (stackman27) [inline x/twap/logic.go:113]: Please let me know if there is preference over using arithmetic operators (+, -, /) vs using sdk math operators (Mul, Quo, Sub)

File: x/twap/logic.go [hunk 2]
```diff
@@ -103,6 +103,19 @@ func (k Keeper) updateRecords(ctx sdk.Context, poolId uint64) error {
...
+
+	if expectedRecordsLength != len(records) {
+		return fmt.Errorf("The number of records do not match")
```
Comment (czarcas7ic) [inline x/twap/logic.go:116]: Can we return expected vs actual length in the error itself?

Comment (stackman27) [reply x/twap/logic.go:116]: Yup!

File: x/twap/logic.go (hunk 1 above)
Comment (ValarDragon) [reply x/twap/logic.go:114]: nah these are native types, should absolutely have been native math :)

File: x/twap/logic.go [hunk 3]
```diff
@@ -103,6 +103,20 @@ func (k Keeper) updateRecords(ctx sdk.Context, poolId uint64) error {
...
 
+	denoms, err := k.ammkeeper.GetPoolDenoms(ctx, poolId)
+	if err != nil {
+		return err
+	}
+
+	// given # of denoms in the pool namely, that for `k` denoms in pool,
+	// there should be k * (k - 1) / 2 records
+	denomNum := len(denoms)
+	expectedRecordsLength := denomNum * (denomNum - 1) / 2
```
Comment (ValarDragon) [inline x/twap/logic.go:114]: Lets add a comment for what we're doing here, or move `expectedRecordsLength := getExpectedNumRecords(ctx, pool Id)` as a separate function?

**Review by ValarDragon — APPROVED** [review]

File: x/twap/logic.go [hunk 4]
```diff
@@ -103,6 +103,20 @@ func (k Keeper) updateRecords(ctx sdk.Context, poolId uint64) error {
 		return err
 	}
 
+	denoms, err := k.ammkeeper.GetPoolDenoms(ctx, poolId)
+	if err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply x/twap/logic.go:108]: Actually, nvm. We will cover this whole function as part of: https://github.com/osmosis-labs/osmosis/issues/2562

**Review by p0mvn — APPROVED** [review]

**Pull request merged**